### PR TITLE
fix(dynamodb): Misconfigured metrics causing empty graphs

### PR DESCRIPTION
**Pull request by iliapolo**
This PR corrects 3 misconfigured metrics we had on the `Table` construct.

### UserErrors

Per the [documentation](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/metrics-dimensions.html) The `table.metricUserErrors()` does not emit the `TableName` dimension. It is actually an account (and region) wide metric.

The fix was to just validate that no dimensions are passed to this operation.

### SystemErrors

Per the [documentation](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/metrics-dimensions.html) The `table.metricSystemErrors()` is always emitted with the `Operation` dimension, and our current implementation does not pass it.

The fix adds an additional `operations` property to the method, that allows passing an array of operations, the returned metric will be a *SUM* over those operations. If no operation is passed, we sum all available operations.

Since the current method returns a `Metric`, returning a math expression won't work since it is an `IMetric` that doesn't extend `Metric`. To avoid breaking changes, we introduce a new method, `metricSystemErrorsForOperations`:

```ts
const totalSystemErrors = table.metricSystemErrorsForOperations();

const getPutSystemErrors = table.metricSystemErrorsForOperations({ 
  operations: [dynamo.Operation.PUT_ITEM, dynamo.Operation.GET_ITEM]
}); 
```

### SuccessfulRequestLatency

Per the [documentation](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/metrics-dimensions.html) The `table.metricSuccessfulRequestLatency()` is always emitted with the `Operation` dimension, and our current implementation does not pass it.

The fix requires user to pass the `Operation` dimension.

So the API is:

```ts
const getLatency = table.metricSuccessfulRequestLatency({ 
  dimensions: {
    Operation: 'GetItem'
  },
});
```

Fixes https://github.com/aws/aws-cdk/issues/11261
Fixes https://github.com/aws/aws-cdk/issues/11269

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:410]: `!` We can't remove this from `ITable`.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:735]:
```suggestion
      throw new Error('only one of `operations` and the Operation dimension can be used');
```

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:752]:
```suggestion
    if (operations.length === 0) {
```

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:791]:
```suggestion
      throw new Error('"dimensions" is not supported for the "UserErrors" metric');
```

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:828]: I'd probably reduce this down quite a bit (in both places); the either/or-ness of them has already been explained just above.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:796]: I know it _shoudn't_ matter, but if we "overridding" here, this should be:

```suggestion
    return this.metric('UserErrors', { statistic: 'sum', ...props, dimensions: {} });
```

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:764]: Should this be using `props.statistic`? What if I want a the system error average, rather than sum?

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:891]: So we're just lucky that there are only 5 read operations, otherwise this wouldn't work? 

I do wonder how useful this metric is on the whole. There can be huge variations between a GET_ITEM and SCAN, and having that lumped into a single metric seems less useful. I wonder if dropping these new convenience methods (for now) would be pragmatic, and then have the community help us understand what combination(s) of these operations makes the most sense to be lumped together.

**Review by njlynch — COMMENTED** [review]
Generally, I think we need to do a "deprecate + create new methods" pattern here, over breaking changes to a stable module. There are users for whom the API wasn't broken, who will hit compilation issues when updating to this solution.

That's likely going to lead to some awkward method naming, but I don't see how we get around that otherwise.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (rix0rrr) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:743]: Not too keen on you changing my simple, one-dimensional metric into a `MathExpression`, just because.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (rix0rrr) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:760]: Can we move all the code around creating a MathExpression on multiple operations into a helper function? I find it very distracting here, and doesn't it occur in multiple places?

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (rix0rrr) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:837]: !!!!!!!!!!!!!!!!!!! ARGLBARGL

Comment (rix0rrr) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:837]: I know this is semi-out of scope but do you think you could see your way to adding a validation to `cloudwatch.Alarm` to that effect as well? Fairly sure this is going to come up more often and I'd rather not have everyone need to figure this out again and again.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (rix0rrr) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:891]: yes

**Review by rix0rrr — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:410]: 👍

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:735]: 👍

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:752]: 👍

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:764]: Its only the default, gets overridden by the user if specified using the `...props` just below.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:853]: This is a little inefficient, iterating twice over the same array. But makes the code much more readable in my opinion, and i'm perfectly happy to sacrifice one for the other.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts]: I love that that you remind me of stuff I wanted to do and forgot :)

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:748]: 5 minutes right? It was a mistake before, unless i'm missing something.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:863]: Any suggestion for additional context here to help users figure out exactly which call is causing this? Just add the method name to the message?

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:921]: Not sure there is a use case for this, but I think this is the *correct* behavior, allowing users to pass additional dimensions that we are not aware of.

**Comment by iliapolo** [comment]
@njlynch @rix0rrr Still no tests but I addressed all your comments. Have a look?

**Comment by iliapolo** [comment]
Just for fun, here is how a combined latency graph will look like:

Screen Shot 2020-11-05 at 2 42 37 PM

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts]: Added to alarm. I also removed this validation because creating graphs doesn't have this limitation. So the `Alarm` id definitely the correct place for it.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:748]: Yeap, verified.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (rix0rrr) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts]: I'm not convinced that `N * (do 2 operations)` is slower than `2 * (do N operations)`

File: packages/@aws-cdk/aws-cloudwatch/test/test.alarm.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-cloudwatch/test/test.alarm.ts:14]:
```suggestion
  'alarm does not accept a math expression with more than 10 metrics'(test: Test) {
```

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:61]: `~` (Super-nitpicky) Totally stylistic, but for these kinds of enums with redundant docstrings, I'd prefer to inline to reduce wasted space.

```suggestion
  /** GetItem */
  GET_ITEM = 'GetItem',
  /** BatchGetItem */
  BATCH_GET_ITEM = 'BatchGetItem',
```

File: packages/@aws-cdk/aws-cloudwatch/lib/alarm.ts
Comment (rix0rrr) [inline packages/@aws-cdk/aws-cloudwatch/lib/alarm.ts:355]: Thiiiiis... is probably not good enough. Is it really 10 submetrics per math expression? Or is it a limit on the total amount of non-expression metrics that an Alarm is going to retrieve?

Any one of the individual submetrics may itself be a mathexpression that contains yet more submetrics, so you may have a bunch of mathexpressions, each with < 10 submetrics, which render to a list of > 10.

They get rendered to a flat list somewhere at the top. The proper validation is probably to just check the length of that flat list (which I imagine is the same as what the service itself is doing).

----

I see that the documentation says that the current implementation is correct, but I wonder if the docs are simplifying language (to not blow people's heads) past the point of correctness...

Comment (rix0rrr) [reply packages/@aws-cdk/aws-cloudwatch/lib/alarm.ts:355]: If you don't feel like experimenting, I guess we leave it like this until someone complains that it doesn't work? :/

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (rix0rrr) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:754]: :(.

This doesn't jibe with all other metrics which add in `TableName` automatically (which is like half of the point of the CDK). Can you not add in TableName automatically like before and just expect users to pass `Operation` ?

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (njlynch) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:863]: Sure. 
```suggestion
      throw new Error("the 'Operation' dimension is not supported for 'metricSuccessfulRequestLatencyForOperations'; use the 'operations' property.");
```

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (rix0rrr) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:754]: (I know it makes the dimension override behavior a bit wonky... but not TOO much)

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (rix0rrr) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:776]: If this metric does not depend on the table object, it should have been a `static` method.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (rix0rrr) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:820]: `Object.values(Operation)` => are you sure this works? I was under the impression that TS enums rendered to bidirectional maps, so I'd expect to see both enum keys and enum values in this list.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (rix0rrr) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:828]: Sum of errors across all operations?

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:923]:
```suggestion
      const operationMetricName = mapper(operation);
```

Right?

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (rix0rrr) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts]: Meh. The stack trace will tell them.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:772]: Just to ensure I understand the perspective here. A user who used these methods without specifying both of the above would never have a valid metric. So while the synth/deploy would succeed, it would be with a broken metric/alarm/dashboard/etc. The mentality is that this is broken -- despite the successful synth -- and so it's now acceptable for a user's synth to fail to warn them of this fact. Correct?

**Review by njlynch — COMMENTED** [review]
Looks good to me. A buncha nit-picky stuff. Needs to be marked as ready for review + merge from master so you can fix the build failures.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:772]: Yeap

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (rix0rrr) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:863]: I still have my doubts about this construction. I think it's bad practice.

First of all:

* I don't see the point of averaging the latency of `GetItem` requests (expected time <10ms) with the latency of `Scan` requests (expected time: ~couple 100ms). What kind of Alarm would you configure in a real-life scenario?
* I'm not even sure if taking an average of averages makes mathematical sense, but I'm 100% sure that in case the user configures a percentile statistic an average of percentiles makes absolutely no sense.

It might make sense to put all the latencies side-by-side in a single dashboard graph (something this method doesn't help with right now), but I'm not sure it makes sense to collapse all these latencies into a single metric.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (rix0rrr) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:901]: Yeah I suppose that's fine.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:863]: @rix0rrr This method can be used with a single operation, in which case it reduces to just the value of the individual metric. I agree combining `Scan` and `GetItem` doesn't make much sense, but what about `Scan` and `Query`? 

In some cases it might not exactly provide something meaningful, but in some cases it might, the user has all the power. Should i maybe just rename to `averageLatencyForOperation`? 

I don't mind getting rid of it - you think it does more harm than good?

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:776]: yes, but that would be breaking.

File: packages/@aws-cdk/aws-cloudwatch/lib/alarm.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-cloudwatch/lib/alarm.ts:355]: I think that eventually it is indeed the metrics that the alarm is going to retrieve. Remember though that people can also do `SEARCH(something that retrieves 100 metrics)` - so we don't really have a way of validating that in the CDK.

Since the current validation definitely makes things a little better, preventing a faulty deployment, i'll leave it at that for now, even though we might do better.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:754]: I considered it, but honestly since this method is deprecated, i think its ok to not invest effort in it.
Users can and should use `table.metricSystemErrorsForOperations({ operations: [GET_ITEM]})` - which doesn't require the table name.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:820]: I saw this on the intrewebs. Tests would have failed if it didn't work

https://github.com/microsoft/TypeScript/issues/17198#issuecomment-448095764

Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:820]: Hmm, now that comment seems funky.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:813]: Strictly speaking the expression shouldn't refer to the metric itself, but to the expression.
`Sum` should appear on the graph or alarm title.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [inline packages/@aws-cdk/aws-dynamodb/lib/table.ts:771]: Since we don't have the `ForOperations` for this metric, i will make it so users don't have to pass the table name.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts]: negligible is an understatement

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts]: Stack traces only show up in verbose...not the best experience. I added the method name.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts]: Removed this metric.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (rix0rrr) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:751]: Which is we would do add new method/`@deprecated`.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (rix0rrr) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:800]: In that comment they do an additional filter though.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (rix0rrr) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:808]: Not sure that's true. A label should describe the timeseries, a graph title describes the complete graph (a title would be "errors vs. latencies" or something, or "call volume and error rate")

**Review by rix0rrr — APPROVED** [review]
Some things to think about for your consideration. Won't block the PR on it, but do be sure you're satisfied with your own choices before merging :).

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:800]:
```console
 ❯ ts-node                                                                                                 [23:44:57]
> enum Op {
... GET = 'GetItem',
... }
undefined
> Object.values(Op)
[ 'GetItem' ]
> 
```

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts:808]: Ok, agreed.

File: packages/@aws-cdk/aws-dynamodb/lib/table.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-dynamodb/lib/table.ts]: Supported now. Users don't need to pass in the TableName

**Pull request merged**
